### DISCUSSION TITLE
No fdkaac support after upgrade

### DISCUSSION THREAD

**Discussion by ambrozieputna** [savonet/liquidsoap#4261] [Installation]
Hi,

Recently I upgraded from Ubuntu 20.4 to 24.4.1 (jumping through 22.4 in between). Liquidsoap 1.4.3 was installed using opam.
After the upgrade, the script fails on start because of a fdkaac stream with the messages: 
```
2024/12/09 16:33:42 [clock.wallclock_main:2] Error when starting output icecast-48: Invalid configuration!
2024/12/09 16:33:42 [clock.wallclock_main:3] Raised by primitive operation at file "src/fdkaac.ml", line 220, characters 4-19
2024/12/09 16:33:42 [clock.wallclock_main:3] Called from file "list.ml", line 110, characters 12-15
2024/12/09 16:33:42 [clock.wallclock_main:3] Called from file "encoder/fdkaac_encoder.ml", line 97, characters 4-49
2024/12/09 16:33:42 [clock.wallclock_main:3] Called from file "encoder/fdkaac_encoder.ml", line 101, characters 14-32
2024/12/09 16:33:42 [clock.wallclock_main:3] Called from file "outputs/icecast2.ml", line 543, characters 22-54
2024/12/09 16:33:42 [clock.wallclock_main:3] Called from file "tools/start_stop.ml", line 108, characters 8-18
2024/12/09 16:33:42 [clock.wallclock_main:3] Called from file "clock.ml", line 245, characters 20-38
```

I made a new opam switch with ocaml 5.3, and manged to install liquidsoap 2.3.0 with fdkaac 0.3.3.
Now I get another error messages:
`2024/12/12 13:26:37 [icecast-48:2] Error while sending data: Lang.Runtime_error { kind: "encoder", msg: "Invalid configuration: aot", pos: [at /home/ambrozie/liquidsoap/scripts/base.liq, line 96, char 15-138] }!`

In the script, the fdkaac encoder is defined as:
`output.icecast(%fdkaac(channels=2, samplerate=44100, bandwidth="auto", bitrate=64, aot="mpeg4_he_aac_v2", transmux="adts", sbr_mode=false),id="icecast-48",`

Please tell me what I'm doing wrong and how to solve this problem.

Ambrozie

**Comment by gAlleb** [comment]
Hi! Ubuntu 24.04 switched from fdk-aac to fdk-aac-free, which has no support for aac-he.

P.S. For `2.3.0` — 4.14.2 compiler is a recommended option for now.

**Comment by ambrozieputna** [comment]
Hi,
Thanks for the quick answer.

I am not very familiar with opam, but what has the fdk-aac-free Ubuntu library to do with the fdkaac opam library?
Indeed, if I do not use a he option for the aot, the script starts just fine.

Is there any way I could use the library with `he` support in my situation?

Ambrozie

**Reply by vitoyucepi** [reply]
Hi @ambrozieputna,
The opam package contains an adapter from the native library to ocaml. `libfdk-aac2` provides a native library, but it was built without features related to `he-aac`.

<details>
<summary>Changelog</summary>

https://changelogs.ubuntu.com/changelogs/pool/universe/f/fdk-aac-free/fdk-aac-free_2.0.2-3~ubuntu4/changelog

> fdk-aac-free (2.0.2-3) UNRELEASED; urgency=medium
>
>  * Package Fedora's stripped version of fdk-aac using the same source
>    package name Fedora uses. This avoids various concerns with the unstripped
>    version. The stripped version should be sufficient for Debian's needs and
>    this allows this library to be accepted into main instead of non-free.
>    (Closes: #981285)

</details>

So all users of `libfdk-aac2`, such as FFmpeg and `ocaml-fdkaac`, cannot use `he-aac` features in order.

I recommend using liquidsoap in [the container](https://hub.docker.com/r/savonet/liquidsoap), but you may want to build your own image.

**Reply by gAlleb** [reply]
> what has the fdk-aac-free Ubuntu library to do with the fdkaac opam library?

The `fdkaac opam library` you are referring to is ocaml binding for the fdk-aac library It builds upon. OCaml fdkaac which you install via opam (https://opam.ocaml.org/packages/fdkaac/) is made by Liquidsoap Team - https://github.com/savonet/ocaml-fdkaac

> Is there any way I could use the library with he support in my situation? 

You can use docker container.

**Reply by gAlleb** [reply]
second. 😃

**Comment by TheRealMattLear** [comment]
Could fdk-aac be compiled manually here to provide support for he-aac?
